Allow clients to send a dummy PSK in ClientHelloOuter (#414)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

   constructed as described below.
1. The value of `ECHConfig.contents.public_name` MUST be placed in the
   "server_name" extension.
1. When the client offers the "pre_shared_key" extension in ClientHelloInner, it
   SHOULD also include a GREASE "pre_shared_key" extension in ClientHelloOuter,
   generated in the manner described in {{grease-psk}}. The client MUST NOT use
   this extension to advertise a PSK to the client-facing server. (See

>    {{flow-clienthello-malleability}}.) When it does, the client also MUST copy
>    the "psk_key_exchange_modes" from the ClientHelloInner into the
>    ClientHelloOuter.

Reviewer: ```suggestion
   {{flow-clienthello-malleability}}.) When the client includes a GREASE "pre_shared_key" extension, 
   it MUST also copy the "psk_key_exchange_modes" from the ClientHelloInner into the ClientHelloOuter.
```

Backend servers (as described in {{server-behavior}}) MUST support the
"ech_is_inner" extension.

### GREASE PSK {#grease-psk}

When offering ECH, the client is not permitted to advertise PSK identities in

> the ClientHelloOuter. However, it is useful to allow the client to add a fake
> "pre_shared_key" extension to the ClientHelloOuter whenever it advertises PSK
> identities in the ClientHelloInner. This provides cover traffic for session
> resumption with the backend server. When resuming a session with the client, the
> backend server sends a "pre_shared_key" extension in its ServerHello. This would

Reviewer: ```suggestion
the ClientHelloOuter. However, the client can send a "pre_shared_key" extension in the
ClientHelloInner. In this case, when resuming a session with the client, the backend 
server sends a "pre_shared_key" extension in its ServerHello. This would
```

Reviewer: (proposed simplification to focus on only extension solicitation violation, rather than on cover traffic, which is the main motivation for this change.)

Reviewer 2: +1. #408 would be another example of a non-cover-traffic motivation which is actually about the real server and not network observers. Worrying about cover traffic suggests needing a threat model and such. The threat model for this is especially weird because the cover traffic can be distinguished with an active client probe:

1. Observe a ClientHello with (possibly GREASE) ECH and (possibly GREASE) PSK.
2. Call the client's bluff and construct a ServerHello that attempts to resume the PSK.
3. See if the client rejects the message.

Author: Done.

appear to a network observer as if the the server were sending this extension
without solicitation, which would violate the extension rules described in

> {{RFC8446}}. Sending a fake "pre_shared_key" extension in the ClientHelloOuter

Reviewer: ```suggestion
{{RFC8446}}. Sending a GREASE "pre_shared_key" extension in the ClientHelloOuter
```

makes it appear to the network as if the extension were negotiated properly.

The client generates the extension payload by constructing an `OfferedPsks`

> structure (see {{RFC8446}}, Section 4.2.11) as follows. For each PSK identity

Reviewer 2: As @kaduk pointed out to me recently, TLS 1.3 actually forbids sending pre_shared_keys without psk_key_exchange_modes and even mandates the server enforce this. If we're going to start patching in individual fields, we probably need to say that too.
https://tools.ietf.org/html/rfc8446#section-4.2.9

Reviewer: 🤦 sigh, yeah.

Author: Yeah, I noticed that too. I guess I sort of thought this would be implied, but being explicit about it seems like a good idea.

Author: The client now MUST copy "psk_key_exchange_modes" from CHInner into CHOuter.

advertised in the ClientHelloInner, the client generates a random PSK identity

> with the same length. It also generates a random, 32-bit, unsinged integer to

Reviewer 2: s/unsinged/unsigned/

> with the same length. It also generates a random, 32-bit, unsinged integer to

Author: Sure, if you like that kind of thing. I prefer my integers to be unsinged.

use as the `obfuscated_ticket_age`. Likewise, for each inner PSK binder, the

> client generates random string of the same length.

Reviewer 2: Suppose the client does this, and then the server sends a ServerHello that claims to handshake with this PSK against ClientHelloOuter. What happens?

(This probably should be an error, i.e. the client should act as if it didn't offer PSKs. But it's ambiguous enough that we should write it down.)

Reviewer: Good point -- I think noting this as an error and failing the handshake seems good enough.

Author: Added this line below:

> If the server replies with a "pre_shared_key" extension in its SeverHello, then
> the client MUST abort the handshake with a "handshake_failure" alert.

Reviewer 2: Minor nitpick: Probably "unsupported_extension" or "illegal_parameter". "handshake_failure" is specifically if everyone followed the protocol correctly, just you were unable to pick a common set of parameter. Whereas the other alerts are variations of "you sent something invalid".

Author: I'll go with "illegal_parameter" :)
